File: api/schemas/reports.py
import datetime
from decimal import Decimal
from enum import Enum

from dateutil import relativedelta
from pydantic import BaseModel
from pydantic import Field

from api.schemas.category import CategoryWithExpenses


class PeriodTypes(Enum):
    day = "day"
    week = "week"
    month = "month"

    def get_relative_delta(self):
        if self == PeriodTypes.day:
            return relativedelta.relativedelta(days=1)
        elif self == PeriodTypes.week:
            return relativedelta.relativedelta(weeks=1)
        elif self == PeriodTypes.month:
            return relativedelta.relativedelta(months=1)


class SpendInPeriodSchema(BaseModel):
    period: datetime.date
    amount: Decimal = Decimal(0)


class NetWorthReport(BaseModel):
    data: list[SpendInPeriodSchema]
    period_type: PeriodTypes = PeriodTypes.day


class PeriodCategoryExpenses(BaseModel):
    period_start: datetime.date
    period_end: datetime.date
    category_expenses: list[CategoryWithExpenses]
    total_expenses: Decimal = Field(default=Decimal(0))

File: frontend/src/interfaces/reports.ts
import { Category } from "src/redux/features/categories/categoriesSlice";

export enum ReportTypes {
  netWorth = "net_worth",
  expenses = "expenses",
}

export enum ReportPeriodType {
  day = "day",
  week = "week",
  month = "month",
}

export interface NetWorthInPeriod {
  period: string;
  amount: number;
}

export interface NetWorthReportResponse {
  data: NetWorthInPeriod[];
  period_type: ReportPeriodType;
}

export interface CategoryWithExpenses extends Category {
  expenses: number;
  percentage_in_period: number;
}

export interface ExpensesReportResponse {
  period_start: string;
  period_end: string;
  category_expenses: CategoryWithExpenses[];
  total_expenses: number;
}

export interface ReportProps {
  startDate: Date;
  endDate: Date;
  periodType?: ReportPeriodType;
}
